### ISSUE TITLE
console twinkle is killed after 5 seconds by SIGALRM

### ISSUE THREAD

**Issue by nahuel** [LubosD/twinkle#89]
I can't use console twinkle because is killed after 5 seconds by SIGALRM, the profile seems to work ok when using the GUI:

```
nahuel@serge:/tmp$ time twinkle -f 1107 -c

Firewall/NAT discovery in progress.
Please wait.

Twinkle 1.10.0, July 15, 2016
Copyright (C) 2005-2015  Michel de Boer and contributors

Users:
* 1107
    1107 <sip:1107@nahtest>

Local IP:       255.255.255.255


1107: registering phone...

Twinkle> 
1107: registration succeeded (expires = 3599 seconds)
Registrar Date: Tue, 29 Nov 2016 19:28:17 GMT


Alarm clock

real	0m5.071s
user	0m0.036s
sys	0m0.024s
nahuel@serge:/tmp$ 
```

strace last part:

```
rt_sigprocmask(SIG_BLOCK, [HUP INT QUIT ALRM TERM TSTP TTIN TTOU], [INT ALRM TERM CHLD], 8) = 0
rt_sigaction(SIGINT, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, {SIG_DFL, [], 0}, 8) = 0
rt_sigaction(SIGTERM, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, {SIG_DFL, [], 0}, 8) = 0
rt_sigaction(SIGHUP, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, {SIG_DFL, [], 0}, 8) = 0
rt_sigaction(SIGQUIT, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, {SIG_DFL, [], 0}, 8) = 0
rt_sigaction(SIGALRM, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, {SIG_DFL, [], 0}, 8) = 0
rt_sigaction(SIGTSTP, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, {SIG_DFL, [], 0}, 8) = 0
rt_sigaction(SIGTTOU, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, {SIG_DFL, [], 0}, 8) = 0
rt_sigaction(SIGTTIN, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, {SIG_DFL, [], 0}, 8) = 0
rt_sigprocmask(SIG_SETMASK, [INT ALRM TERM CHLD], NULL, 8) = 0
rt_sigaction(SIGWINCH, {0x7fd361164400, [], SA_RESTORER|SA_RESTART, 0x7fd361b8d670}, {SIG_DFL, [], 0}, 8) = 0
write(1, "Twinkle> ", 9Twinkle> )                = 9
pselect6(1, [0], NULL, NULL, NULL, {[], 8}
1107: registration succeeded (expires = 3599 seconds)
Registrar Date: Tue, 29 Nov 2016 19:30:44 GMT


) = ? ERESTARTNOHAND (To be restarted if no handler)
--- SIGALRM {si_signo=SIGALRM, si_code=SI_KERNEL} ---
rt_sigreturn({mask=[INT ALRM TERM CHLD]}) = -1 EINTR (Interrupted system call)
ioctl(1, TCXONC, TCOON)                 = 0
ioctl(0, TCGETS, {B38400 opost isig -icanon -echo ...}) = 0
ioctl(0, SNDCTL_TMR_STOP or TCSETSW, {B38400 opost isig icanon echo ...}) = 0
ioctl(0, TCGETS, {B38400 opost isig icanon echo ...}) = 0
rt_sigaction(SIGINT, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGTERM, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGHUP, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGQUIT, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGALRM, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGTSTP, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGTTOU, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGTTIN, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGWINCH, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, {0x7fd361164400, [], SA_RESTORER|SA_RESTART, 0x7fd361b8d670}, 8) = 0
rt_sigprocmask(SIG_BLOCK, NULL, [INT ALRM TERM CHLD], 8) = 0
kill(10344, SIGALRM)                    = 0
rt_sigprocmask(SIG_SETMASK, [INT TERM CHLD], NULL, 8) = 0
ioctl(0, TIOCGWINSZ, {ws_row=13, ws_col=194, ws_xpixel=0, ws_ypixel=0}) = 0
ioctl(0, TIOCSWINSZ, {ws_row=13, ws_col=194, ws_xpixel=0, ws_ypixel=0}) = 0
ioctl(0, TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(0, TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(0, SNDCTL_TMR_STOP or TCSETSW, {B38400 opost isig -icanon -echo ...}) = 0
ioctl(0, TCGETS, {B38400 opost isig -icanon -echo ...}) = 0
rt_sigprocmask(SIG_BLOCK, [HUP INT QUIT ALRM TERM TSTP TTIN TTOU], [INT TERM CHLD], 8) = 0
rt_sigaction(SIGINT, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGTERM, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGHUP, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGQUIT, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGALRM, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGTSTP, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGTTOU, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGTTIN, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigprocmask(SIG_SETMASK, [INT TERM CHLD], NULL, 8) = 0
rt_sigaction(SIGWINCH, {0x7fd361164400, [], SA_RESTORER|SA_RESTART, 0x7fd361b8d670}, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
pselect6(1, [0], NULL, NULL, NULL, {[], 8}) = ? ERESTARTNOHAND (To be restarted if no handler)
--- SIGALRM {si_signo=SIGALRM, si_code=SI_KERNEL} ---
rt_sigreturn({mask=[INT TERM CHLD]})    = -1 EINTR (Interrupted system call)
ioctl(1, TCXONC, TCOON)                 = 0
ioctl(0, TCGETS, {B38400 opost isig -icanon -echo ...}) = 0
ioctl(0, SNDCTL_TMR_STOP or TCSETSW, {B38400 opost isig icanon echo ...}) = 0
ioctl(0, TCGETS, {B38400 opost isig icanon echo ...}) = 0
rt_sigaction(SIGINT, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGTERM, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGHUP, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGQUIT, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGALRM, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGTSTP, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGTTOU, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGTTIN, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, {0x7fd361164ec0, [], SA_RESTORER, 0x7fd361b8d670}, 8) = 0
rt_sigaction(SIGWINCH, {SIG_DFL, [], SA_RESTORER, 0x7fd361b8d670}, {0x7fd361164400, [], SA_RESTORER|SA_RESTART, 0x7fd361b8d670}, 8) = 0
rt_sigprocmask(SIG_BLOCK, NULL, [INT TERM CHLD], 8) = 0
kill(10344, SIGALRM)                    = 0
--- SIGALRM {si_signo=SIGALRM, si_code=SI_USER, si_pid=10344, si_uid=1000} ---
+++ killed by SIGALRM +++
Alarm clock
```

Running on Ubuntu 16.10, twinkle installed from ubuntu repo, version: Twinkle 1.10.0 - 15 July 2016

**Comment by scheibler** [comment]
Same result with twinkle-console version 1:1.10.1+dfsg-2 on a 64 bit Debian Stretch (testing).

**Comment by fbriere** [comment]
This is caused by readline 7.0, most probably due to the following changelog entry:

> Readline's input handler now performs signal processing if read(2) is interrupted by SIGALRM or SIGVTALRM.

Upon calling `readline()`, SIGALRM is apparently unblocked, leading to this behavior.  (For the record, disabling `rl_catch_signals` doesn't appear to have any effect.)

This is all a bit beyond my expertise, so I can't tell whether or not this is an actual bug with readline.  (It certainly looks like it, though.)  I'll try sending a bug report upstream, and see what happens.

**Comment by fbriere** [comment]
Email sent to the bug-readline mailing-list: https://lists.gnu.org/archive/html/bug-readline/2017-01/msg00002.html

**Comment by petercolberg** [comment]
Thanks to @fbriere's debugging effort, this issue has been resolved [upstream](https://ftp.gnu.org/gnu/readline/readline-7.0-patches/readline70-003) and in [Debian](https://bugs.debian.org/852750) stretch.

**Comment by fbriere** [comment]
Given that stretch has been released more than six months ago, I think it's safe to close this issue.  :smile:

**Comment by purpleidea** [comment]
This happens with the Fedora-32 package.

**Comment by fbriere** [comment]
> This happens with the Fedora-32 package.

According to rpmfind.net, Fedora 32 shipped with libreadline 8.0, so this would be quite puzzling.  Could you provide us with a [backtrace](https://github.com/LubosD/twinkle/wiki/Generating-a-backtrace) of this happening?

**Comment by purpleidea** [comment]
> Could you provide us with a [backtrace](https://github.com/LubosD/twinkle/wiki/Generating-a-backtrace) of this happening?

I commented on this a year ago. I'm no longer on F32.
